### PR TITLE
*: Add variable `tidb_placement_mode` and remove `placement_checks`

### LINKED ISSUE #31092: Add `tidb_placement_mode`

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #31092

### What is changed and how it works?

Introduce new variable `tidb_placement_mode` and it has two enums:

- strict: The default value, check all placement settings in ddl operations
- ignore: Ignore all placement settings in ddl operations

BR will use `CreateTableWithInfo` , `CreateSchemaWithInfo` to create table and database, so we must only valid placement policy in these methods. This PR moved `checkAndNormalizePlacement` from out into these methods.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
1. Remove sys variable `placement_checks`
2. Introduce new system variable `tidb_placement_mode`
```

### CODE REVIEW

File: sessionctx/variable/sysvar.go [hunk 1]
```diff
@@ -856,6 +856,16 @@ var defaultSysVars = []*SysVar{
 		s.EnableAlterPlacement = TiDBOptOn(val)
 		return nil
 	}},
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBPlacementMode, Value: DefTiDBPlacementMode, Type: TypeStr, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		mode := PlacementMode(normalizedValue)
+		if !mode.Valid() {
+			return normalizedValue, ErrWrongTypeForVar.GenWithStackByArgs(TiDBPlacementMode)
+		}
+		return string(mode), nil
```
Comment (morgo) [inline sessionctx/variable/sysvar.go:864]: You can actually remove the validation func completely here - just change the type to `TypeEnum` with the permitted values for strict/ignore.

**Comment by morgo** [comment]
Can you remove the sysvar `PLACEMENT_CHECKS` in the same PR? This will help reduce the confusion.

File: sessionctx/variable/sysvar.go (hunk 1 above)
Comment (lcwangchao) [reply sessionctx/variable/sysvar.go:864]: Sorry, I'm not very understand. Could you explain it about change the type to `TypeEnum` ?

Comment (morgo) [reply sessionctx/variable/sysvar.go:864]: See https://github.com/pingcap/tidb/blob/fb03dacbcaa11145003838d672a6956ac60d7ff6/sessionctx/variable/sysvar.go#L606-L609 for an example.

The `TypeEnum` validates that options belong to the list of `PossibleValues`, and will return an error if they are invalid. It is also case insensitive and normalizes to the case in the `PossibleValues`.

(We have many such cases for enumerated options, which is why using the built-in helper is useful).

Comment (lcwangchao) [reply sessionctx/variable/sysvar.go:864]: Okay. Thank you

File: ddl/partition.go [hunk 2]
```diff
@@ -471,16 +480,28 @@ func setPartitionPlacementFromOptions(partition *model.PartitionDefinition, opti
 				return err
 			}
 		case ast.TableOptionPlacementPolicy:
+			hasPlacement = true
+			if ignorePlacement {
+				hasPlacement = true
```
Comment (djshow832) [inline ddl/partition.go:485]: Remove this line.

File: ddl/partition.go [hunk 3]
```diff
@@ -471,16 +480,28 @@ func setPartitionPlacementFromOptions(partition *model.PartitionDefinition, opti
...
+
 			partition.PlacementPolicyRef = &model.PolicyRefInfo{
 				Name: model.NewCIStr(opt.StrValue),
 			}
 		}
 	}
 
+	if ignorePlacement && hasPlacement {
+		ctx.GetSessionVars().StmtCtx.AppendNote(errors.New(
+			fmt.Sprintf("Placement options is ignored when TIDB_PLACEMENT_MODE is '%s'", variable.PlacementModeIgnore),
```
Comment (djshow832) [inline ddl/partition.go:497]:
```suggestion
			fmt.Sprintf("Placement options are ignored when TIDB_PLACEMENT_MODE is '%s'", variable.PlacementModeIgnore),
```

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -2032,6 +2040,13 @@ func (d *ddl) CreateTableWithInfo(
 		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(dbName)
 	}
 
+	sessVars := ctx.GetSessionVars()
+	if sessVars.PlacementMode == variable.PlacementModeIgnore && removeTablePlacement(tbInfo) {
+		sessVars.StmtCtx.AppendNote(errors.New(
+			fmt.Sprintf("Placement options is ignored when TIDB_PLACEMENT_MODE is '%s'", variable.PlacementModeIgnore),
```
Comment (djshow832) [inline ddl/ddl_api.go:2046]:
```suggestion
			fmt.Sprintf("Placement options are ignored when TIDB_PLACEMENT_MODE is '%s'", variable.PlacementModeIgnore),
```

**Review by djshow832 — COMMENTED** [review]
Rest LGTM

**Comment by lcwangchao** [comment]
> Can you remove the sysvar `PLACEMENT_CHECKS` in the same PR? This will help reduce the confusion.

`PLACEMENT_CHECKS` removed, PTAL again @morgo

File: docs/design/2020-06-24-placement-rules-in-sql.md [hunk 5]
```diff
@@ -458,18 +458,16 @@ When placement policies are specified, they should be validated for correctness:
...
-SET PLACEMENT_CHECKS=0;
+The system variable `TIDB_PLACEMENT_MODE` can control whether to ignore placement options in ddl or not. It has two values:
 
-CREATE TABLE t3 (a int) PLACEMENT POLICY `mycompanypolicy`;
-```
+* STRICT: The default value. It means that we should check all placement settings strictly in ddl operations.
+* IGNORE: It means tidb should ignore all placement settings in ddl operations.
 
-If a table is imported when `PLACEMENT_CHECKS` is `OFF`, and the placement policy does not validate, then the same rules of fallback apply as in the case `DROP PLACEMENT POLICY` (where policy is still in use).
+If a table is imported when `TIDB_PLACEMENT_MODE` is `IGNORE`, it is equivalent to importing a table without any placement options.
```
Comment (xhebox) [inline docs/design/2020-06-24-placement-rules-in-sql.md:470]: Morgan has included the document changes in #31423, so could you merge your changes to that PR?

**Review by xhebox — COMMENTED** [review]
LGTM

File: docs/design/2020-06-24-placement-rules-in-sql.md (hunk 5 above)
Comment (lcwangchao) [reply docs/design/2020-06-24-placement-rules-in-sql.md:470]: Maybe I should revert the changes in document and left it to morgon

Comment (xhebox) [reply docs/design/2020-06-24-placement-rules-in-sql.md:470]: Maybe, I am OK to both.

**Review by xhebox — APPROVED** [review]

**Review by morgo — APPROVED** [review]

**Pull request merged**